state/application.go

```diff
@@ -2230,3 +2236,129 @@ func (a *Application) PasswordValid(password string) bool {
 	}
 	return false
 }
+
+// UnitUpdateProperties holds information used to update
+// the state model for the unit.
+type UnitUpdateProperties struct {
+	ProviderId string
+	Status     *status.StatusInfo
+}
+
+// UpdateUnits applies the given application unit update operations.
+func (a *Application) UpdateUnits(unitsOp *UpdateUnitsOperation) error {
+	return a.st.ApplyOperation(unitsOp)
+}
+
+// UpdateUnitsOperation is a model operation for updating
+// some units of an application.
+type UpdateUnitsOperation struct {
+	Adds    []*AddUnitOperation
+	Deletes []*DestroyUnitOperation
+	Updates []*UpdateUnitOperation
+}
+
+func (op *UpdateUnitsOperation) allOps() []ModelOperation {
+	var all []ModelOperation
+	for _, mop := range op.Adds {
+		all = append(all, mop)
+	}
+	for _, mop := range op.Deletes {
+		all = append(all, mop)
+	}
+	for _, mop := range op.Updates {
+		all = append(all, mop)
+	}
+	return all
+}
+
+// Build is part of the ModelOperation interface.
+func (op *UpdateUnitsOperation) Build(attempt int) ([]txn.Op, error) {
+	var ops []txn.Op
+
+	all := op.allOps()
+	for _, op := range all {
+		switch nextOps, err := op.Build(attempt); err {
+		case jujutxn.ErrNoOperations:
+			continue
+		case nil:
+			ops = append(ops, nextOps...)
+		default:
+			return nil, errors.Trace(err)
+		}
+	}
+	return ops, nil
+}
+
+// Done is part of the ModelOperation interface.
+func (op *UpdateUnitsOperation) Done(err error) error {
+	if err != nil {
+		return errors.Annotate(err, "updating units")
+	}
+	all := op.allOps()
+	for _, op := range all {
+		if err := op.Done(nil); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// AddOperation returns a model operation that will add a unit.
+func (a *Application) AddOperation(props UnitUpdateProperties) *AddUnitOperation {
+	return &AddUnitOperation{
+		application: &Application{st: a.st, doc: a.doc},
+		props:       props,
+	}
+}
+
+// AddUnitOperation is a model operation that will add a unit.
+type AddUnitOperation struct {
+	application *Application
+	props       UnitUpdateProperties
+
+	unitName string
+}
+
+// Build is part of the ModelOperation interface.
+func (op *AddUnitOperation) Build(attempt int) ([]txn.Op, error) {
+	var ops []txn.Op
+
+	addUnitArgs := AddUnitParams{
+		ProviderId: op.props.ProviderId,
+	}
+	name, addOps, err := op.application.addUnitOps("", addUnitArgs, nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	op.unitName = name
+	ops = append(ops, addOps...)
+
+	return ops, nil
+}
+
+// Done is part of the ModelOperation interface.
+func (op *AddUnitOperation) Done(err error) error {
+	if err != nil {
+		return errors.Annotatef(err, "adding unit to %q", op.application.Name())
+	}
+	// We do a separate status update here because we require all units to be
+	// created as "allocationing". If the add operation specificies a status,
+	// that status is used to update the initial "allocating" status. We then
+	// get the expected 2 status entries in history.
```

Comment:
something here about this being in a separate transaction, but that's OK because a failure will cause an update in the future when the relevant worker restarts

Reply:
done